io-uring-test/src/tests/queue.rs

```diff
@@ -198,3 +198,95 @@ pub fn test_msg_ring_data<S: squeue::EntryMarker, C: cqueue::EntryMarker>(
+) -> anyhow::Result<()> {
+    use std::os::unix::io::AsRawFd;
+
+    // This test requires both `IORING_OP_MSG_RING` opcode and `IORING_MSG_SEND_FD` flag.
+    // We directly probe for the opcode, but the flag only appeared later
+    // (in kernel 6.0) and cannot be probed.
+    // Thus, we additionaly check for the `IORING_OP_SEND_ZC` opcode (>= 6.0)
+    // as a proxy instead.
+    require!(
+        test;
+        test.probe.is_supported(opcode::MsgRingData::CODE);
+    );
+    require!(
+        test;
+        test.probe.is_supported(opcode::SendZc::CODE);
+    );
+
+    println!("test msg_ring_send_fd");
+
+    // Ensure there are no fixed files on the ring, so that slots allocation
+    // will start at 0.
+    let _ = ring.submitter().unregister_files();
+
+    // Create a new destination ring.
+    let mut temp_ring = IoUring::new(8)?;
+
+    // As of kernel 6.1, FD sending only works if the receiving ring already
+    // has some fixed files registered.
+    // Thus, we create two tempfiles and register one to each ring (at slot 0).
+    {
+        let tmp1 = tempfile::tempfile()?;
+        ring.submitter()
+            .register_files(&[tmp1.as_raw_fd()])
+            .unwrap();
+        let tmp2 = tempfile::tempfile()?;
+        temp_ring
+            .submitter()
+            .register_files(&[tmp2.as_raw_fd()])
+            .unwrap();
+    }
+
+    // Send the fixed file from the source ring (on slot 0)
+    // to the temp ring (to slot 1).
+    // This will generate two completion events, one on each ring.
+    unsafe {
+        let fd = types::Fd(temp_ring.as_raw_fd());
+        ring.submission()
+            .push(
+                &opcode::MsgRingSendFd::new(fd, types::Fixed(0), types::Fixed(1), 11, 22)
+                    .build()
+                    .into(),
+            )
+            .expect("queue is full");
+    }
+    ring.submit_and_wait(1)?;
+    {
+        let source_cqes: Vec<cqueue::Entry> = ring.completion().map(Into::into).collect();
+        assert_eq!(source_cqes.len(), 1);
+        assert_eq!(source_cqes[0].user_data(), 0);
+        assert_eq!(source_cqes[0].result(), 0);
+        assert_eq!(source_cqes[0].flags(), 0);
+
+        let dest_cqes: Vec<cqueue::Entry> = temp_ring.completion().map(Into::into).collect();
+        assert_eq!(dest_cqes.len(), 1);
+        assert_eq!(dest_cqes[0].user_data(), 22);
+        assert_eq!(dest_cqes[0].result(), 11);
+        assert_eq!(dest_cqes[0].flags(), 0);
+    }
+
+    // NOTE(lucab): we stop testing here, as there seem to be some kernel glitches (observed
+    // on v6.1).
+    // Looking at the `temp_ring` fdinfo, it seems that the FD is correctly received but
+    // wrongly installed to slot 0. So the rings status at this point is confusing:
+    // ```
+    // [/proc/<PID>/fdinfo/<source ring>]
+    // UserFiles:      1
+    //     0: .tmpy3CuuJ
+    // [/proc/<PID>/fdinfo/<dest ring>]
+    // UserFiles:      1
+    //     0: .tmpy3CuuJ
+    // ```
```

Comment:
@axboe I'm confused by what I'm observing here on the receiving ring. Am I somehow misusing the kernel API?